File: swift/src/Sodium/Sodium/Sodium/SodiumTests/TestCommon.swift
//
//  TestCommon.swift
//  Sodium
//
//  Created by Andrew Bradnan on 5/7/16.
//  Copyright © 2016 Whirlygig Ventures. All rights reserved.
//

import XCTest
@testable import Sodium

extension SodiumTests {

    func test_Base_send1() {
        let s = Transaction.run { StreamSink<String>() }!
        var out = [String]()
        let l = Transaction.run { s.listen{ out.append($0) } }!
        Transaction.runVoid{
            s.send("a")
        }
        Transaction.runVoid{
            s.send("b")
        }
        
        l.unlisten()

        XCTAssert(["a","b"] == out, "testSendStream() failed \(out)")
    }

    
    func test_Operational_split() {
        let a = Transaction.noThrowRun{ StreamSink<[String]>() }
        let b = Transaction.noThrowRun { _ in Operational.split(a) }
        var b_0 = [String]()
        let b_0_l = Transaction.noThrowRun { b.listen{ b_0.append($0) } }
        
        Transaction.runVoid{
            a.send(["a","b"])
        }
        b_0_l.unlisten()

        XCTAssert(["a","b"] == b_0, "testSplit() failed \(b_0)")
    }
    
    func test_Operational_defer1() {
        let a = Transaction.noThrowRun { StreamSink<String>() }
        let b = Transaction.noThrowRun { Operational.Defer(a) }

        var b_0 = [String]()
        let b_0_l = Transaction.noThrowRun{ b.listen{ b_0.append($0) } }
        
        Transaction.runVoid{
            a.send("a")
        }
        b_0_l.unlisten()
        XCTAssert(["a"] == b_0, "testDefer() failed \(b_0)")
        
        var b_1 = [String]()
        let b_1_l = Transaction.noThrowRun{ b.listen{ b_1.append($0) } }

        Transaction.runVoid{ a.send("b") }
        b_1_l.unlisten()
        XCTAssert(["b"] == b_1, "testDefer() failed \(b_1)")
    }
    
    func test_Operational_defer2() {
        let a = Transaction.run{ StreamSink<String>() }!
        let b = Transaction.run{ StreamSink<String>() }!
        let c = Transaction.run{ Operational.Defer(a).orElse(b) }!
    
        var c_0 = [String]()
        let c_0_l = Transaction.run{ c.listen{ c_0.append($0) } }!
    
        Transaction.runVoid{ a.send("a") }
        c_0_l.unlisten()
        XCTAssert(["a"] == c_0, "testDefer2() failed \(c_0)")

        var c_1 = [String]()
        let c_1_l = Transaction.run{ c.listen{ c_1.append($0) } }!
        Transaction.runVoid{
            a.send("b")
            b.send("B")
        }
        c_1_l.unlisten()
        XCTAssert(["B", "b"] == c_1, "testDefer2() failed \(c_1)")
    }

    func test_Stream_orElse1() {
        let a = Transaction.run{ StreamSink<Int>() }!
        let b = Transaction.run{ StreamSink<Int>() }!
        let c = Transaction.run{ a.orElse(b) }!
        var c_0 = [Int]()
        let c_0_l = Transaction.run{ c.listen{ c_0.append($0) } }!
    
        Transaction.runVoid{ a.send(0) }
        c_0_l.unlisten()
        XCTAssert([0] == c_0, "testOrElse1() failed \(c_0)")
    
        var c_1 = [Int]()
        let c_1_l = Transaction.run{ c.listen{ c_1.append($0) } }!
        Transaction.runVoid{ b.send(10) }
        c_1_l.unlisten()
        XCTAssert([10] == c_1, "testOrElse1() failed \(c_1)")
        var c_2 = [Int]()
        let c_2_l = Transaction.run{ c.listen{ c_2.append($0) } }!
        Transaction.runVoid{
            a.send(2)
            b.send(20)
        }
        c_2_l.unlisten()
        XCTAssert([2] == c_2, "testOrElse1() failed \(c_2)")
        
        var c_3 = [Int]()
        let c_3_l = Transaction.run{ c.listen{ c_3.append($0) } }!
        Transaction.runVoid{ b.send(30) }
        c_3_l.unlisten()
        XCTAssert([30] == c_3, "testOrElse1() failed \(c_3)")
    }

    /*

    
    
    public void test_Operational_deferSimultaneous() {
    StreamSink<String> a = Transaction.run{
    StreamSink<String> a_ = new StreamSink()
    return a_
    })
    StreamSink<String> b = Transaction.run{
    StreamSink<String> b_ = new StreamSink()
    return b_
    })
    Stream<String> c = Transaction.run{
    Stream<String> c_ = Operational.defer(a).orElse(Operational.defer(b))
    return c_
    })
    List<String> c_0 = new ArrayList<>()
    Listener c_0_l = Transaction.run{
    Listener c_0_l_ = c.listen{
    c_0.add(val)
    })
    return c_0_l_
    })
    Transaction.runVoid{
    b.send("A")
    })
    c_0_l.unlisten()
     XCTAssert(["A"] == c_0, "testDeferSimultaneous() failed \(c_0)")
     //assertEquals(Arrays.<String>asList("A"),c_0)
    List<String> c_1 = new ArrayList<>()
    Listener c_1_l = Transaction.run{
    Listener c_1_l_ = c.listen{
    c_1.add(val)
    })
    return c_1_l_
    })
    Transaction.runVoid{
    a.send("b")
    b.send("B")
    })
    c_1_l.unlisten()
     XCTAssert(["b"] == c_1, "testDeferSimultaneous() failed \(c_1)")
    //assertEquals(Arrays.<String>asList("b"),c_1)
    }
*/
}

File: python/tests/test_common.py
from typing import List

from sodiumfrp import operational
from sodiumfrp import StreamSink, Transaction

def test_base_send1() -> None:
    s: StreamSink[str] = Transaction.run(StreamSink)
    out: List[str] = []
    l = Transaction.run(lambda: s.listen(out.append))
    Transaction.run(lambda: s.send("a"))
    Transaction.run(lambda: s.send("b"))
    l.unlisten()
    assert ["a", "b"] == out

def test_operational_split() -> None:
    a: StreamSink[List[str]] = Transaction.run(StreamSink)
    b = Transaction.run(lambda: operational.split(a))
    b_0: List[str] = []
    b_0_l = Transaction.run(lambda: b.listen(b_0.append))
    Transaction.run(lambda: a.send(["a", "b"]))
    b_0_l.unlisten()
    assert ["a", "b"] == b_0

def test_operational_defer1() -> None:
    a: StreamSink[str] = Transaction.run(StreamSink)
    b = Transaction.run(lambda: operational.defer(a))
    b_0: List[str] = []
    b_0_l = Transaction.run(lambda: b.listen(b_0.append))
    Transaction.run(lambda: a.send("a"))
    b_0_l.unlisten()
    assert ["a"] == b_0
    b_1: List[str] = []
    b_1_l = Transaction.run(lambda: b.listen(b_1.append))
    Transaction.run(lambda: a.send("b"))
    b_1_l.unlisten()
    assert ["b"] == b_1

def test_operational_defer2() -> None:
    a: StreamSink[str] = Transaction.run(StreamSink)
    b: StreamSink[str] = Transaction.run(StreamSink)
    c = Transaction.run(lambda: operational.defer(a).or_else(b))
    c_0: List[str] = []
    c_0_l = Transaction.run(lambda: c.listen(c_0.append))
    Transaction.run(lambda: a.send("a"))
    c_0_l.unlisten()
    assert ["a"] == c_0
    c_1: List[str] = []
    c_1_l = Transaction.run(lambda: c.listen(c_1.append))
    def transaction() -> None:
        a.send("b")
        b.send("B")
    Transaction.run(transaction)
    c_1_l.unlisten()
    assert ["B", "b"] == c_1

def test_stream_or_else1() -> None:
    a: StreamSink[int] = Transaction.run(StreamSink)
    b: StreamSink[int] = Transaction.run(StreamSink)
    c = Transaction.run(lambda: a.or_else(b))
    c_0: List[int] = []
    c_0_l = Transaction.run(lambda: c.listen(c_0.append))
    Transaction.run(lambda: a.send(0))
    c_0_l.unlisten()
    assert [0] == c_0
    c_1: List[int] = []
    c_1_l = Transaction.run(lambda: c.listen(c_1.append))
    Transaction.run(lambda: b.send(10))
    c_1_l.unlisten()
    assert [10] == c_1
    c_2: List[int] = []
    c_2_l = Transaction.run(lambda: c.listen(c_2.append))
    def transaction() -> None:
        a.send(2)
        b.send(20)
    Transaction.run(transaction)
    c_2_l.unlisten()
    assert [2] == c_2
    c_3: List[int] = []
    c_3_l = Transaction.run(lambda: c.listen(c_3.append))
    Transaction.run(lambda: b.send(30))
    c_3_l.unlisten()
    assert [30] == c_3

def test_operational_defer_simultaneous() -> None:
    a: StreamSink[str] = Transaction.run(StreamSink)
    b: StreamSink[str] = Transaction.run(StreamSink)
    c = Transaction.run(
        lambda: operational.defer(a).or_else(operational.defer(b)))
    c_0: List[str] = []
    c_0_l = Transaction.run(lambda: c.listen(c_0.append))
    Transaction.run(lambda: b.send("A"))
    c_0_l.unlisten()
    assert ["A"] == c_0
    c_1: List[str] = []
    c_1_l = Transaction.run(lambda: c.listen(c_1.append))
    def transaction() -> None:
        a.send("b")
        b.send("B")
    Transaction.run(transaction)
    c_1_l.unlisten()
    assert ["b"] == c_1
